File: Top-Interview-Questions-Leetcode/Easy-Collections/linked-list/symmetric-tree.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def __isSymmetric_recursively(self, node1: Optional[TreeNode], node2: Optional[TreeNode]) -> bool:
        if not node1 and not node2:
            return True
        elif not node1 or not node2:
            return False
        elif node1.val != node2.val:
            return False
        else:
            return self.__isSymmetric_recursively(node1.left, node2.right) & self.__isSymmetric_recursively(node1.right, node2.left)
    
    def __isSymmetric_iteratively(self, root: Optional[TreeNode]) -> bool:
        nodes = [[root.left, root.right]]
        
        while nodes:
            left, right = nodes[-1]
            nodes.pop()
            
            if not left and not right:
                continue
            elif not left or not right:
                return False
            elif left.val != right.val:
                return False
            else:
                nodes.append([left.left, right.right])
                nodes.append([left.right, right.left])
        
        return True
    
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        # return self.__isSymmetric_recursively(root.left, root.right)
        return self.__isSymmetric_iteratively(root)
        

File: Top-Interview-Questions-Leetcode/Easy-Collections/linked-list/symmetric-tree.cpp
#include <iostream>
#include <queue>
#include <utility>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
private:
    
    bool isSymmetric_recursively(TreeNode* left, TreeNode* right) {
        if (!left and !right) return true;
        if (left or right) return false;
        if (left->val != right->val) return false;
        
        return isSymmetric_recursively(left->left, right->right) & isSymmetric_recursively(left->right, right->left);
    }
    
    bool isSymmetric_iteratively(TreeNode* root) {
        std::queue<std::pair<TreeNode*, TreeNode*> > q;
        q.push({root->left, root->right});
        
        while (!q.empty()) {
            auto [left, right] = q.front();
            q.pop();
            
            if (!left and !right) continue;
            else if (!left or !right) return false;
            else if (left->val != right->val) return false;
            else {
                q.push({left->left, right->right});
                q.push({left->right, right->left});
            }
        }
        
        return true;
    }
    
public:
    bool isSymmetric(TreeNode* root) {
        // return isSymmetric_recursively(root->left, root->right);
        return isSymmetric_iteratively(root);
    }
};